wdl-format/src/element.rs

```diff
@@ -0,0 +1,508 @@
+
+        if let Some(children) = self.children() {
+            for child in children {
+                results
+                    .entry(child.element().kind())
+                    .or_insert(Vec::new())
+                    // NOTE: this clone is very cheap, as the underlying
+                    // elements are mostly reference counts.
+                    .push(child.to_owned())
+            }
+        }
+
+        results
+    }
+
+    /// Writes any preceeding trivia to the stream.
+    pub fn write_preceeding_trivia(&self, stream: &mut TokenStream<PreToken>) {
+        if let Some(trivia) = self.trivia().preceeding() {
+            for t in trivia.filter(|t| !matches!(t.element().kind(), SyntaxKind::Whitespace)) {
+                t.write(stream);
+            }
+        }
+    }
+
+    /// Writes any inline trivia to the stream.
+    pub fn write_inline_trivia(&self, stream: &mut TokenStream<PreToken>) {
+        if let Some(trivia) = self.trivia().inline() {
+            for t in trivia.filter(|t| !matches!(t.element().kind(), SyntaxKind::Whitespace)) {
+                t.write(stream);
+            }
+        }
+    }
+}
+
+/// An extension trait for formatting [`Element`]s.
+pub trait AstElementFormatExt {
+    /// Consumes `self` and returns the [`Element`] as a [`FormatElement`].
+    fn into_format_element(self) -> FormatElement;
+}
+
+impl AstElementFormatExt for Element {
+    fn into_format_element(self) -> FormatElement
+    where
+        Self: Sized,
+    {
+        let children = match &self {
+            Element::Node(node) => collate(node),
+            Element::Token(_) => None,
+        };
+
+        FormatElement::new(self, Default::default(), children)
+    }
+}
+
+/// Collects a list of iterables into an [`Option<NonEmpty>`].
+fn collect_optional<T>(mut iter: impl Iterator<Item = T>) -> Option<NonEmpty<T>> {
+    if let Some(first) = iter.next() {
+        let mut vec = NonEmpty::new(first);
+        vec.extend(iter);
+        Some(vec)
+    } else {
+        None
+    }
+}
+
+/// Takes elements while a particular predicate is true _without_ consuming the
+/// element that breaks the chain.
+fn take_while_peek<'a, I, P>(
+    iter: &'a mut Peekable<I>,
+    predicate: P,
+) -> impl Iterator<Item = I::Item> + 'a
+where
+    I: Iterator,
+    P: Fn(&I::Item) -> bool + 'a,
+{
+    std::iter::from_fn(move || {
+        if let Some(next_item) = iter.peek() {
+            if predicate(next_item) {
+                iter.next()
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    })
+}
+
+/// Collates the children of a particular node.
+fn collate(node: &Node) -> Option<NonEmpty<Box<FormatElement>>> {
+    let mut results = Vec::new();
+    let mut stream = node
+        .syntax()
+        .children_with_tokens()
+        .map(Element::cast)
+        .peekable();
+
+    while stream.peek().is_some() {
+        let preceeding = collect_optional(
```

Comment:
```suggestion
        let preceding = collect_optional(
```
Please do a project-wide find+replace for `preceeding -> preceding`. I'll stop pointing these out now.